python/private/whl_target_platforms.bzl

```diff
@@ -118,68 +103,129 @@ def _whl_priority(value):
-    for platform_tag, whl in candidates.items():
-        if platform_tag == "any":
+        if want_abis and parsed.abi_tag not in want_abis:
+            # Filter out incompatible ABIs
             continue
 
-        if "musl" in platform_tag:
-            # Ignore musl wheels for now
+        if version_limit != -1 and whl_version_min > version_limit:
+            continue
+
+        compatible = False
+        if parsed.platform_tag == "any":
+            compatible = True
+        else:
+            for p in whl_target_platforms(parsed.platform_tag):
+                if p.target_platform in want_platforms:
+                    compatible = True
+                    break
+
+        if not compatible:
             continue
 
-        platform_tag = ".".join({_LEGACY_ALIASES.get(p, p): True for p in platform_tag.split(".")})
-        platforms = whl_target_platforms(platform_tag)
-        for p in platforms:
-            target_plats.setdefault("{}_{}".format(p.os, p.cpu), []).append(platform_tag)
+        for implementation in supported_implementations:
+            candidates.setdefault(
+                (
+                    parsed.abi_tag,
+                    parsed.platform_tag,
+                ),
+                {},
+            ).setdefault(
+                (
+                    # prefer cp implementation
+                    implementation == "cp",
+                    # prefer higher versions
+                    whl_version_min,
+                    # prefer abi3 over none
+                    parsed.abi_tag != "none",
+                    # prefer cpx abi over abi3
+                    parsed.abi_tag != "abi3",
+                ),
+                [],
+            ).append(whl)
+
+    return [
+        candidates[key][sorted(v)[-1]][-1]
+        for key, v in candidates.items()
+    ]
+
+def select_whl(*, whls, want_platform):
+    """Select a suitable wheel from a list.
 
-    for p, platform_tags in target_plats.items():
-        if has_any:
-            platform_tags.append("any")
+    Args:
+        whls(list[struct]): A list of candidates.
+        want_platform(str): The target platform.
 
-        target_plats[p] = sorted(platform_tags, key = _whl_priority)
+    Returns:
+        None or a struct with `url`, `sha256` and `filename` attributes for the
+        selected whl. If no match is found, None is returned.
+    """
+    if not whls:
+        return None
 
-    want = target_plats.get(want_platform)
-    if not want:
-        return want
+    # TODO @aignas 2024-05-23: once we do the selection in the hub repo using
+    # an actual select, then this function will be the one that is used within
+    # the repository context instead of `select_whl`.
+    whls = select_whls(
+        whls = whls,
+        want_version = "",
+        want_platforms = [want_platform],
+    )
+
+    candidates = {
+        parse_whl_name(w.filename).platform_tag: w
+        for w in whls
+        if "musllinux_" not in w.filename
```

Comment:
I'm guessing the musllinux gets special cased because we don't have way to model this yet?